### PR TITLE
Replace click with argparse

### LINKED ISSUE #137: Straighten out CLI usage
This is something I noticed while working on #110 

A) It is pretty convenient to do a two step deposit by first running "hermes" and then "hermes deposit". Maybe it is not necessary to have "hermes --deposit", "hermes --postprocess" etc. This might also help the code structure, as it's kind of hard to integrate Click options and arguments with these flags, as it requires manual parsing instead of relying on Click decorator mechanics. I did also encounter some errors about unknown options etc.

B) I noticed that the help is showing the option to have another command after the first:
```
Usage: hermes deposit [OPTIONS] COMMAND [ARGS]...
```
I'm not sure this is on purpose to concat commands or a bug because of our special handling of commands.

I definitely ran into a problem with this. When trying to get file artifacts going as arguments, running `hermes --deposit --auth-token XXX --file YYY`, resulted in a "command not known" error.

### PR DESCRIPTION
**Pull request by sdruskat**
This PR removes click from the direct dependencies of hermes, and re-implements the CLI functionality with `argparse`. Multi-commands a.k.a. subcommand chaining is removed.

- Fixes #137 

- [ ] Supersede ADR10 with argparse ADR and no multi-commands

### CODE REVIEW

**Comment by led02** [comment]
**question:** While working on this, should I also handle #182 (which will break the API and will require adaptations in docs @SKernchen and probably workflows @poikilotherm)

**Comment by led02** [comment]
**swear:** Obviously, somebody ran an auto formatter on this branch, forcing me to manually merge tons of commits with uncountable lines where only the quotation marks were changed. But hidden in between are changes that come from the Pydantic Settings change. 😡

File: src/hermes/commands/workflow.py [hunk 1]
```diff
@@ -207,27 +221,28 @@ def deposit(click_ctx: click.Context, initial, auth_token, file):
...
-        deposit_plugin()
-    except (RuntimeError, MisconfigurationError) as e:
-        _log.error(f"Error in {plugin_group!r} plugin {plugin_name!r}: {e}")
-        click_ctx.exit(1)
+    for entry_point in loaded_entry_points:
+        try:
+            deposit_plugin(path, config_path, initial, auth_token, files, ctx)
+        except (RuntimeError, MisconfigurationError) as e:
+            _log.error(f"Error in {group!r} entry point {name!r}: {e}")
+            sys.exit(1)
```
Comment (zyzzyxdonta) [inline src/hermes/commands/workflow.py:237]: There's something wrong here. `loaded_entry_points` does not exist and neither do `group` and `name`.

**Comment by led02** [comment]
> I like where this is going 👍🏻 However, it doesn't seem to be wired up correctly yet, as none of the commands do anything so far.

Focus was on the cli.py until now. The design of the command line interface should be kind of stable, already started to do some magic Pydantic Settings stuff here. Unfortunatly, SWIM ran a code formatter and pushed the results, hence there are some more changes indicated that there really are... :(

Still need to:

- wire up the implementation
- finish the dynamic settings assembly
- make the other phases also use plug-in classes

File: src/hermes/commands/workflow.py [hunk 2]
```diff
@@ -7,13 +7,14 @@
 # SPDX-FileContributor: David Pape
 # SPDX-FileContributor: Oliver Bertuch
 
+import argparse
```
[the remark this answers is not in the document]
Comment (led02) [reply src/hermes/commands/workflow.py:10]: The `workflow.py` will be gone once this PR is ready.

File: src/hermes/commands/deposit/base.py [hunk 3]
```diff
@@ -21,8 +19,7 @@ class BaseDepositPlugin(HermesPlugin):
     TODO: describe workflow... needs refactoring to be less stateful!
     """
 
-    def __init__(self, click_ctx: click.Context, ctx: CodeMetaContext) -> None:
-        self.click_ctx = click_ctx
+    def __init__(self, ctx):
         self.ctx = ctx
```
Comment (led02) [inline src/hermes/commands/deposit/base.py:23]: **thought:** We might decide to keep the ``HermesContext`` and derived mischief for now. I see the need to refactor the "deposit" (base) plugin which is heavily state dependent. However, we are doing data processing, hence I think it's a good idea to reflect that in the design...

However, to go forward I think it's easiest to keep the ctx stuff for now, design the plugin interface in a way that only the "Command" does interact with it and have a better hinge point for later refactoring.

To sum it up: For now, do the refactoring using context, design new plugins that the data is visible, later get rid of contexts.
Disclaimer: The purpose of this note is to derive an issue out of it ;)

File: src/hermes/commands/deposit/invenio.py [hunk 4]
```diff
@@ -252,14 +252,14 @@ class InvenioDepositPlugin(BaseDepositPlugin):
...
-        super().__init__(click_ctx, ctx)
+    def __init__(self, ctx: CodeMetaContext, client=None, resolver=None) -> None:
+        super().__init__(ctx)
 
         self.invenio_context_path = ContextPath.parse(f"deposit.{self.platform_name}")
         self.invenio_ctx = None
 
         if client is None:
-            auth_token = self.click_ctx.params.get("auth_token")
+            auth_token = self.ctx.params.get("auth_token")
```
Comment (led02) [inline src/hermes/commands/deposit/invenio.py:262]: **chore:** We need to decide on how to pass command line arguments to the plugins. Currently, the only thing that plugins can interact with the "outside" world is the settings. Hence, I'd opt for putting everything that is needed by the plugin in to a setting. You can also configure those settings from command line by using e.g,, ``hermes harvest -O invenio.auth_tokten MY_AUTH_TOKEN -O invenio.initial True``

File: src/hermes/commands/process/base.py [hunk 5]
```diff
@@ -17,6 +21,11 @@ class HermesProcessCommand(HermesCommand):
     """ Process the collected metadata into a common dataset. """
 
     command_name = "process"
+    settings_class = None
 
     def __call__(self, args: argparse.Namespace) -> None:
-        pass
+        self.args = args
+        # TODO: get harvested data
+        # TODO: Merge Datasets
```
Comment (led02) [inline src/hermes/commands/process/base.py:29]: Also:
- Call pre-processors plugins (well, not really clear)
- Call processor plugins

These interfaces are not at all well-defined by now. We had different entry points, but I don't see this suitable for the new CLI without introducing new commands.

I cloud image that **processor** plugins have a pre-process stage (that is called before the merge) and a process stage (that is called after the merge). As we do not have lots to process for now, it's also fine to work without plugins here for now.

**Comment by SKernchen** [comment]
- Fixes #156 
The processor will prioritize the first harvester in the toml.
